presqt/osf/classes/storage_folder.py

```diff
@@ -59,6 +75,188 @@ def get_all_files(self):
+        kind : str
+            The kind of resource we are attempting to get.
+        klass :str
+            The resource's kind associated class
+        """
+        children = self._follow_next(url)
+
+        while children:
+            child = children.pop()
+            kind_ = child['attributes']['kind']
+            if kind_ == kind:
+                yield klass(child, self.session)
+
+    def get_folder_by_name(self, folder_name):
+        """
+        Gets a folder object based on the name. Only looks for top level folders.
+
+        Parameters
+        ----------
+        folder_name : str
+            Name of the folder we want to find within the container.
+
+        Returns
+        -------
+        Returns an instance of the requested Folder class.
+        """
+        for folder in self.iter_children(self._files_url, 'folder', Folder):
+            if folder.title == folder_name:
+                return folder
+
+    def get_file_by_name(self, file_name):
+        """
+        Gets a file object based on the name. Only looks for top level files.
+
+        Parameters
+        ----------
+        file_name : str
+            Name of the file we want to find within the container.
+
+        Returns
+        -------
+        Returns an instance of the requested File class.
+        """
+        for file in self.iter_children(self._files_url, 'file', File):
+            if file.title == file_name:
+                return file
+
+    def create_folder(self, folder_name):
+        """
+        Create a new sub-folder for this container.
+
+        Parameters
+        ----------
+        folder_name : str
+            Name of the folder to create.
+
+        Returns
+        -------
+        Class instance of the created folder.
+        """
+        response = self.put(self._new_folder_url, params={'name': folder_name})
+        if response.status_code == 409:
+            return self.get_folder_by_name(folder_name)
+
+        elif response.status_code == 201:
+            return self.get_folder_by_name(folder_name)
+
+        else:
+            raise PresQTResponseException(
+                "Response has status code {} while creating folder {}".format(response.status_code,
+                                                                              folder_name),
+                status.HTTP_400_BAD_REQUEST)
+
+    def create_file(self, file_name, file_to_write, file_duplicate_action):
+        """
+        Upload a file to a container.
+
+        Parameters
+        ----------
+        file_name : str
+            Name of the file to create.
+        file_to_write : binary file
+            File to create.
+        file_duplicate_action : str
+            Flag for how to handle the case of the file already existing.
+
+        Returns
+        -------
+        Class instance of the created file.
+        """
+        # When uploading a large file (>a few MB) that already exists
```

Comment:
Hmmm.... makes me worry a bit about other connection errors being treated as duplicated.
Not sure of a solution, just a concern.